enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,358 @@
+   the Hybrid overlay CIDR. So that we can reuse the cluster network in the
+   migration.
+4. We need to modify OVN-K to allow modifying the hybrid overlay CIDR on the fly.
+
+On the OpenShift SDN side, when a node is converted to OVN Kubernetes,
+it shall be almost transparent to the control-plane. But we need to introduce a
+'migration mode' for OpenShift SDN by:
+
+1. Change ingress NetworkPolicy processing to be based entirely on pod IPs
+   rather than using namespace VNIDs, since packets from ovn nodes will have
+   the VNID 0 set.
+2. To be compatible with Windows node VXLAN implementation, OVN-K hybrid overlay
+   uses the host interface MAC as VXLAN inner MAC. When packets arrive at the
+   br0 of the SDN node, they cannot be forwarded to the pod interface, due to
+   the MAC mismatch. We need to add flows for each pod which swaps the dst MAC
+   to the pod interface mac.
+
+### The Traffic Path
+
+#### Packets going from OpenShift SDN to OVN Kubernetes
+
+On the SDN side, it doesn't need to know if the peer node is a SDN node or a OVN
+node. We reuse the existing VXLAN tunnel rules one the SDN side.
+- Egress NetworkPolicy rules and service proxying happen as normal
+- When the packet reaches table 90, it will hit a “send via vxlan” rule that was
+  generated based on a HostSubnet object.
+
+On the OVN side:
+- OVN accepts the packet via the VXLAN tunnel, ignore the VNID set by SDN and
+  then just routes it normally.
+- Ingress NetworkPolicy processing will happen when the packet reaches the
+  destination pod’s switch port, just like normal.
+  - Our NetworkPolicy rules are all based on IP addresses, not “logical input
+    port”, etc, so it doesn’t matter that the packets came from outside OVN and
+    have no useful OVN metadata.
+
+#### Packets going from OVN Kubernetes to OpenShift SDN
+
+On the ovn side:
+- The packet just follow the same path of hybrid overlay. The packet just has to
+  get routed out the VXLAN tunnel with VNID 0.
+
+On the sdn side:
+- We have to change ingress NetworkPolicy processing to be based entirely on pod
+  IPs rather than using namespace VNIDs, since packets from ovn nodes won’t have
+  the VNID set. There is already code to generate the rules that way though,
+  because egress NetworkPolicy already works that way.
+
+### The Migration Process
+
+#### Migration Setup
+
+1. The admin kicks off the migration process by editing the CNO configuration
+
+2. CNO will label each existing node with, eg:
+
+    ```yaml
+    migration.network.openshift.io/plugin: OpenShiftSDN
+    k8s.ovn.org/hybrid-overlay-node-subnet: 10.129.0.0/23
+    ```
+
+   (CNO knows how to figure out the host subnet by looking at openshift-sdn
+   HostSubnet objects. When migrating in the reverse direction, it knows to look
+   at the ovn-kubernetes annotations.)
+
+3. CNO will redeploy the openshift-sdn DaemonSets, enabling “migration mode”,
+   and adding node affinity to the node DaemonSet:
```

Comment:
What is the impact to the cluster as these updates happen?  If any possible disruption it should be called out in the Risks section above, even if it can be couched as no worse than any other DS rollout.

Reply:
openshift-sdn-node will be restarted on each node. The restart itself is immediate. After restarting, the node process will need to reload its caches, which may take a few seconds in large clusters (eg thousands of pods), during which time it would lag in processing updates to the cluster. (So, eg, a pod launched at this time may take an extra few seconds to become ready.)

In clusters with less than thousands of pods the restart is likely to be completely unnoticeable.